File: timApp/modules/pistelasku/pistelasku.py
"""
TIM example plugin: a pistelaskundrome checker.
"""
import re
from typing import Union

import attr
from flask import jsonify, render_template_string
from marshmallow import Schema, fields, post_load, validates, ValidationError
from marshmallow.utils import missing
from webargs.flaskparser import use_args

from pluginserver_flask import GenericMarkupModel, GenericMarkupSchema, GenericHtmlSchema, GenericHtmlModel, \
    GenericAnswerSchema, GenericAnswerModel, Missing, \
    InfoSchema, create_app


@attr.s(auto_attribs=True)
class PistelaskuStateModel:
    """Model for the information that is stored in TIM database for each answer."""
    userword: str


class PistelaskuStateSchema(Schema):
    userword = fields.Str(required=True)

    @post_load
    def make_obj(self, data):
        return PistelaskuStateModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class PistelaskuMarkupModel(GenericMarkupModel):
    points_array: Union[str, Missing] = missing
    inputstem: Union[str, Missing] = missing
    needed_len: Union[int, Missing] = missing
    initword: Union[str, Missing] = missing
    cols: Union[int, Missing] = missing
    inputplaceholder: Union[str, Missing] = missing


class PistelaskuMarkupSchema(GenericMarkupSchema):
    points_array = fields.List(fields.List(fields.Number()))
    inputstem = fields.Str()
    needed_len = fields.Int()
    initword = fields.Str()
    cols = fields.Int()
    inputplaceholder = fields.Str()

    @validates('points_array')
    def validate_points_array(self, value):
        if len(value) != 2 or not all(len(v) == 2 for v in value):
            raise ValidationError('Must be of size 2 x 2.')

    @post_load
    def make_obj(self, data):
        return PistelaskuMarkupModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class PistelaskuInputModel:
    """Model for the information that is sent from browser (plugin AngularJS component)."""
    userword: str
    pistelaskuOK: bool
    nosave: bool = missing


class PistelaskuInputSchema(Schema):
    userword = fields.Str(required=True)
    pistelaskuOK = fields.Bool(required=True)
    nosave = fields.Bool()

    @validates('userword')
    def validate_userword(self, word):
        if not word:
            raise ValidationError('Must not be empty.')

    @post_load
    def make_obj(self, data):
        return PistelaskuInputModel(**data)


class PistelaskuAttrs(Schema):
    """Common fields for HTML and answer routes."""
    markup = fields.Nested(PistelaskuMarkupSchema)
    state = fields.Nested(PistelaskuStateSchema, allow_none=True, required=True)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class PistelaskuHtmlModel(GenericHtmlModel[PistelaskuInputModel, PistelaskuMarkupModel, PistelaskuStateModel]):
    def get_component_html_name(self) -> str:
        return 'pistelasku-runner'

    def get_static_html(self) -> str:
        return render_static_pistelasku(self)

    def get_browser_json(self):
        r = super().get_browser_json()
        if self.state:
            r['userword'] = self.state.userword
        return r

    class Meta:
        strict = True


class PistelaskuHtmlSchema(PistelaskuAttrs, GenericHtmlSchema):
    info = fields.Nested(InfoSchema, allow_none=True, required=True)

    @post_load
    def make_obj(self, data):
        # noinspection PyArgumentList
        return PistelaskuHtmlModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class PistelaskuAnswerModel(GenericAnswerModel[PistelaskuInputModel, PistelaskuMarkupModel, PistelaskuStateModel]):
    pass


class PistelaskuAnswerSchema(PistelaskuAttrs, GenericAnswerSchema):
    input = fields.Nested(PistelaskuInputSchema, required=True)

    @post_load
    def make_obj(self, data):
        # noinspection PyArgumentList
        return PistelaskuAnswerModel(**data)

    class Meta:
        strict = True


def render_static_pistelasku(m: PistelaskuHtmlModel):
    return render_template_string(
        """
<div class="csRunDiv no-popup-menu">
    <h4>{{ header }}</h4>
    <p class="stem">{{ stem }}</p>
    <div><label>{{ inputstem or '' }} <span>
        <input type="text"
               class="form-control"
               placeholder="{{inputplaceholder}}"
               size="{{cols}}"></span></label>
    </div>
    <button class="timButton">
        {{ buttonText or button or "Save" }}
    </button>
    <a>{{ resetText }}</a>
    <p class="plgfooter">{{ footer }}</p>
</div>
        """,
        **attr.asdict(m.markup),
    )


app = create_app(__name__, PistelaskuHtmlSchema())


@app.route('/answer/', methods=['put'])
@use_args(PistelaskuAnswerSchema(strict=True), locations=("json",))
def answer(args: PistelaskuAnswerModel):
    web = {}
    result = {'web': web}
    needed_len = args.markup.needed_len
    userword = args.input.userword
    pistelasku_ok = args.input.pistelaskuOK
    len_ok = True
    if needed_len:
        len_ok = check_letters(userword, needed_len)
    if not len_ok:
        web['error'] = "Wrong length"
    if not needed_len and not pistelasku_ok:
        len_ok = False
    points_array = args.markup.points_array or [[0, 0.25], [0.5, 1]]
    points = points_array[pistelasku_ok][len_ok]

    # plugin can ask not to save the word
    nosave = args.input.nosave
    if not nosave:
        tim_info = {"points": points}
        save = {"userword": userword}
        result["save"] = save
        result["tim_info"] = tim_info
        web['result'] = "saved"

    return jsonify(result)


def check_letters(word: str, needed_len: int) -> bool:
    """Checks if word has needed amount of chars.

    :param word: word to check
    :param needed_len: how many letters needed
    :return: true if len match

    """
    s = word.upper()
    return len(re.sub("[^[A-ZÅÄÖ]", "", s)) == needed_len


@app.route('/reqs/')
@app.route('/reqs')
def reqs():
    templates = ["""
``` {#ekapistelasku plugin="pistelasku"}
header: Kirjoita pistelasku
stem: Kirjoita pistelasku, jossa on 5 kirjainta.
-points_array: [[0, 0.1], [0.6, 1]]
inputstem: "Pistelaskusi:"
needed_len: 5
answerLimit: 3
initword: muikku
cols: 20
```""", """
``` {#tokapistelasku plugin="pistelasku"}
header: Kirjoita pistelasku
stem: Kirjoita pistelasku, jossa on 7 kirjainta.
-points_array: [[0, 0.1], [0.6, 1]]
inputstem: "Pistelaskusi:"
needed_len: 7
answerLimit: 4
initword: muikku
cols: 20
```"""]
    return jsonify({
        "js": ["js/build/pistelasku.js"],
        "multihtml": True,
        "css": ["css/pistelasku.css"],
        'editor_tabs': [
            {
                'text': 'Plugins',
                'items': [
                    {
                        'text': 'Pistelasku',
                        'items': [
                            {
                                'data': templates[0].strip(),
                                'text': '5 letters',
                                'expl': 'Add a 5-letter pistelaskundrome task',
                            },
                            {
                                'data': templates[1].strip(),
                                'text': '7 letters',
                                'expl': 'Add a 7-letter pistelaskundrome task',
                            },
                        ],
                    },
                ],
            },
        ],
    },
    )


if __name__ == '__main__':
    app.run(
        host='0.0.0.0',
        port=5000,
        debug=False,  # for live reloading, this can be turned on
    )

File: timApp/modules/pistelasku/js/pistelasku.ts
/**
 * Defines the client-side implementation of an example plugin (a pistelaskundrome checker).
 */
import angular, {INgModelOptions} from "angular";
import * as t from "io-ts";
import {GenericPluginMarkup, nullable, PluginBase, withDefault} from "tim/plugin/util";
import {$http} from "tim/util/ngimport";
import {to} from "tim/util/utils";
import {valueDefu} from "tim/util/utils";
import {number} from "../../../static/scripts/jspm_packages/npm/io-ts@1.4.1/lib";

const pistelaskuApp = angular.module("pistelaskuApp", ["ngSanitize"]);
export const moduleDefs = [pistelaskuApp];

const PistelaskuMarkup = t.intersection([
    t.partial({
        initword: t.string,
        inputplaceholder: nullable(t.string),
        inputstem: t.string,
        initword2: t.number,
        inputplaceholder2: nullable(t.number),
        inputstem2: t.number,
    }),
    GenericPluginMarkup,
    t.type({
        // all withDefaults should come here; NOT in t.partial
        autoupdate: withDefault(t.number, 500),
        cols: withDefault(t.number, 20),
    }),
]);
const PistelaskuAll = t.intersection([
    t.partial({
        userword: t.string,
        userword2: t.number,
    }),
    t.type({markup: PistelaskuMarkup}),
]);

function isPistelaskundrome(s: string) {
    let sc = s.toLowerCase();
    sc = sc.replace(/[^a-zåöä]/g, "");
    for (let i1 = 0, i2 = sc.length - 1; i1 < i2; i1++, i2--) {
        if (sc[i1] !== sc[i2]) {
            return false;
        }
    }
    return true;
}

class PistelaskuController extends PluginBase<t.TypeOf<typeof PistelaskuMarkup>, t.TypeOf<typeof PistelaskuAll>, typeof PistelaskuAll> {
    private result?: string;
    private error?: string;
    private isRunning = false;
    private userword = "";
    private userword2 = number;
    private runTestGreen = false;
    private modelOpts!: INgModelOptions; // initialized in $onInit, so need to assure TypeScript with "!"

    getDefaultMarkup() {
        return {};
    }

    buttonText() {
        return super.buttonText() || "Save";
    }

    $onInit() {
        super.$onInit();
        this.userword = this.attrsall.userword || this.attrs.initword || "";
        this.modelOpts = {debounce: this.autoupdate};
        this.checkPistelaskundrome();
    }

    get edited() {
        return this.attrs.initword !== this.userword;
    }

    get autoupdate(): number {
        return this.attrs.autoupdate;
    }

    get inputplaceholder() {
        return this.attrs.inputplaceholder || null;
    }

    get inputstem() {
        return this.attrs.inputstem || null;
    }

    get inputstem2() {
        return this.attrs.inputstem2 || null;
    }

    get cols() {
        return this.attrs.cols;
    }

    get resetText() {
        return valueDefu(this.attrs.resetText, "Reset");
    }

    checkPistelaskundrome() {
        const is = isPistelaskundrome(this.userword);
        this.runTestGreen = is;
        return is;
    }

    initCode() {
        this.userword = this.attrs.initword || "";
        this.userword2 = number;
        this.error = undefined;
        this.result = undefined;
        this.checkPistelaskundrome();
    }

    saveText() {
        this.doSaveText(false);
    }

    async doSaveText(nosave: boolean) {
        this.error = "... saving ...";
        this.isRunning = true;
        this.result = undefined;
        const params = {
            input: {
                nosave: false,
                pistelaskuOK: this.checkPistelaskundrome(),
                userword: this.userword,
                userword2: this.userword2,
            },
        };

        if (nosave) {
            params.input.nosave = true;
        }
        const url = this.pluginMeta.getAnswerUrl();
        const r = await to($http.put<{web: {result: string, error?: string}}>(url, params));
        this.isRunning = false;
        if (r.ok) {
            const data = r.result.data;
            this.error = data.web.error;
            this.result = data.web.result;
        } else {
            this.error = "Infinite loop or some other error?";
        }
    }

    protected getAttributeType() {
        return PistelaskuAll;
    }
}

pistelaskuApp.component("pistelaskuRunner", {
    bindings: {
        json: "@",
    },
    controller: PistelaskuController,
    template: `
<div class="csRunDiv no-popup-menu">
    <h4 ng-if="::$ctrl.header" ng-bind-html="::$ctrl.header"></h4>
    <p ng-if="::$ctrl.stem">{{::$ctrl.stem}}</p>
    <div class="form-inline"><label>{{::$ctrl.inputstem}} <span>
        <input type="text"
               class="form-control"
               ng-model="$ctrl.userword"
               ng-model-options="::$ctrl.modelOpts"
               ng-change="$ctrl.checkPistelaskundrome()"
               ng-trim="false"
               placeholder="{{::$ctrl.inputplaceholder}}"
               size="{{::$ctrl.cols}}"></span></label>
        <span class="unitTestGreen" ng-if="$ctrl.runTestGreen && $ctrl.userword">OK</span>
        <span class="unitTestRed" ng-if="!$ctrl.runTestGreen">Wrong</span><br>
    </div>
    <div class="form-inline"><label>{{::$ctrl.inputstem2}} <span>   
        <input type="number"
               class="form-control"
               ng-model="$ctrl.userword2"
               ng-model-options="::$ctrl.modelOpts"
               ng-change="$ctrl.checkPistelaskundrome()"
               ng-trim="false"
               placeholder="{{::$ctrl.inputplaceholder2}}"
               size="{{::$ctrl.cols}}"></span></label>
        <span class="unitTestGreen" ng-if="$ctrl.runTestGreen && $ctrl.userword2">OK</span>
        <span class="unitTestRed" ng-if="!$ctrl.runTestGreen">Wrong</span>
    </div>
    <button class="timButton"
            ng-if="::$ctrl.buttonText()"
            ng-disabled="$ctrl.isRunning || !$ctrl.userword || !$ctrl.userword2"
            ng-click="$ctrl.saveText()">
        {{::$ctrl.buttonText()}}
    </button>
    <a href="" ng-if="$ctrl.edited" ng-click="$ctrl.initCode()">{{::$ctrl.resetText}}</a>
    <div ng-if="$ctrl.error" ng-bind-html="$ctrl.error"></div>
    <pre ng-if="$ctrl.result">{{$ctrl.result}}</pre>
    <p ng-if="::$ctrl.footer" ng-bind="::$ctrl.footer" class="plgfooter"></p>
</div>
`,
});
